Bluetooth: meaningful error codes on advertising start

Qbicz opened the pull request:
Translate HCI error codes to POSIX error codes in order to be able to
distinguish reason of connectable advertising start failure.

Signed-off-by: Filip Kubicz <filip.kubicz@nordicsemi.no>

Qbicz commented:
When trying to start connectable advertising when max number of connections is already established, I will get `BT_HCI_ERR_CONN_LIMIT_EXCEEDED` which translates to `-EMLINK`. This POSIX code is usually used in filesystem context, but for Bluetooth it fits well here in my opinion.

@cvinayak @jhedberg @carlescufi If you see there are more cases that could be handled, feel free to add your suggestions.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -10502,13 +10502,12 @@ u32_t radio_adv_enable(u16_t interval, u8_t chan_map, u8_t filter_policy,
 		link = mem_acquire(&_radio.link_rx_free);
 		if (!link) {
-			return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+			return BT_HCI_ERR_CONN_LIMIT_EXCEEDED;
```
Please swap the order of the two mem_acquire, and use the new error code for the connection context allocation.

jhedberg reviewed (changes requested):
I'm not super happy with 'EMLINK', but I couldn't find any better choice either. Something else related to "out of resources" might have worked as well. So generally this PR looks good to me, however two changes that I think would be good:

 - Split into two commits: one for the controller, another for the host
 - For the host commit add to the documentation of bt_le_adv_start() a note about the special error code, and that it's only guaranteed for the Zephyr controller, whereas with other's you might get EIO.

cvinayak commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -274,7 +274,15 @@ int bt_hci_cmd_send_sync(u16_t opcode, struct net_buf *buf,
-		err = -EIO;
+		switch (cmd(buf)->status) {
+		case BT_HCI_ERR_CONN_LIMIT_EXCEEDED:
+			err = -EMLINK;
```
How about -ECONNREFUSED or -ENOBUFS or -ENOMEM?

jhedberg replied:
I like the first one, since it makes it clear that this is about connections and not advertising as such (which may not otherwise be clear to the user since the API in question is bt_le_adv_start)

Qbicz committed:
Bluetooth: ctlr: Return different error codes on advertising start

Return meaninful HCI error when it's not possible to start advertising
because of maximum number of connections already in use.

Signed-off-by: Filip Kubicz <filip.kubicz@nordicsemi.no>

Qbicz committed:
Bluetooth: host: Meaningful error codes on advertising start

Translate HCI error codes to POSIX error codes in order to be able to
distinguish reason of connectable advertising start failure.

Signed-off-by: Filip Kubicz <filip.kubicz@nordicsemi.no>

Qbicz commented:
@jhedberg @cvinayak I have split the commit, documented the return value and used `-EISCONN`.

cvinayak commented on include/bluetooth/bluetooth.h:
```diff
@@ -343,6 +343,10 @@ struct bt_le_adv_param {
  *
  *  @return Zero on success or (negative) error code otherwise.
+ *  @return -EISCONN when connectable advertising is requested and there
+ *  is already maximum number of connections established. This error
```
indentation, align with the start of the description and capital 'W' in 'When' .

cvinayak reviewed (commented):
@Qbicz When controller supports no connected roles, then -ECONNREFUSED seems better, unless there is already a return of -ENOTSUP somewhere in the code path?

Qbicz commented:
@cvinayak fixed indentation, used -ECONNREFUSED return code.

Qbicz commented:
@cvinayak can you recheck?

cvinayak reviewed (approved):
LGTM

Qbicz commented:
@cvinayak, @jhedberg, can you merge?
